libvirt/domain.go

```diff
@@ -0,0 +1,171 @@
+		return true, domWaitLeaseDone, nil
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"waiting-addresses"},
+		Target:     []string{domWaitLeaseDone},
+		Refresh:    waitFunc,
+		Timeout:    timeout,
+		MinTimeout: 10 * time.Second,
+		Delay:      5 * time.Second,
+	}
+
+	_, err := stateConf.WaitForState()
+	log.Print("[DEBUG] wait-for-leases was successful")
+	return err
+}
+
+func domainIfaceHasAddress(domain libvirt.Domain, iface libvirtxml.DomainInterface) (found bool, ignore bool, err error) {
+
+	mac := strings.ToUpper(iface.MAC.Address)
+	if mac == "" {
+		log.Printf("[DEBUG] Can't wait without a MAC address: ignoring interface %+v.\n", iface)
+		// we can't get the ip without a mac address
+		return false, true, nil
+	}
+
+	log.Printf("[DEBUG] waiting for network address for iface=%s\n", mac)
+	ifacesWithAddr, err := domainGetIfacesInfo(domain)
+	if err != nil {
+		return false, false, fmt.Errorf("Error retrieving interface addresses: %s", err)
+	}
+	log.Printf("[DEBUG] ifaces with addresses: %+v\n", ifacesWithAddr)
+
+	for _, ifaceWithAddr := range ifacesWithAddr {
+		if mac == strings.ToUpper(ifaceWithAddr.Hwaddr) {
+			log.Printf("[DEBUG] found IPs for MAC=%+v: %+v\n", mac, ifaceWithAddr.Addrs)
+			return true, false, nil
+		}
+	}
+
+	log.Printf("[DEBUG] %+v doesn't have IP address(es) yet...\n", mac)
+	return false, false, nil
+}
+
+func domainGetState(domain libvirt.Domain) (string, error) {
+	state, _, err := domain.GetState()
+	if err != nil {
+		return "", err
+	}
+
+	var stateStr string
+
+	switch state {
+	case libvirt.DOMAIN_NOSTATE:
+		stateStr = "nostate"
+	case libvirt.DOMAIN_RUNNING:
+		stateStr = "running"
+	case libvirt.DOMAIN_BLOCKED:
+		stateStr = "blocked"
+	case libvirt.DOMAIN_PAUSED:
+		stateStr = "paused"
+	case libvirt.DOMAIN_SHUTDOWN:
+		stateStr = "shutdown"
+	case libvirt.DOMAIN_CRASHED:
+		stateStr = "crashed"
+	case libvirt.DOMAIN_PMSUSPENDED:
+		stateStr = "pmsuspended"
+	case libvirt.DOMAIN_SHUTOFF:
+		stateStr = "shutoff"
+	default:
+		stateStr = fmt.Sprintf("unknown: %v", state)
+	}
+
+	return stateStr, nil
+}
+
+func domainIsRunning(domain libvirt.Domain) (bool, error) {
+	state, _, err := domain.GetState()
+	if err != nil {
+		return false, fmt.Errorf("Couldn't get state of domain: %s", err)
+	}
+
+	return state == libvirt.DOMAIN_RUNNING, nil
+}
+
+func domainGetIfacesInfo(domain libvirt.Domain) ([]libvirt.DomainInterface, error) {
+
+	// get all the interfaces using the qemu-agent, this includes also
+	// interfaces that are not attached to networks managed by libvirt
+	// (eg. bridges, macvtap,...)
+	log.Print("[DEBUG] fetching networking interfaces using qemu-agent")
+	interfaces := qemuAgentGetInterfacesInfo(&domain, true)
+	if len(interfaces) > 0 {
+		// the agent will always return all the interfaces, both the
+		// ones managed by libvirt and the ones attached to bridge interfaces
+		// or macvtap. Hence it has the highest priority
+		return interfaces, nil
+	}
```

Comment:
The new qemu-agent code causes our tests to time out. They cannot connect to any qemu-agent because there is none. We could work around this issue:
```go
var skipQemuAgent = "false"

// ...

func domainGetIfacesInfo(domain libvirt.Domain) ([]libvirt.DomainInterface, error) {
	if skipQemuAgent == "false" {
		// get all the interfaces using the qemu-agent, this includes also
		// interfaces that are not attached to networks managed by libvirt
		// (eg. bridges, macvtap,...)
		log.Print("[DEBUG] fetching networking interfaces using qemu-agent")
		interfaces := qemuAgentGetInterfacesInfo(&domain, true)
		if len(interfaces) > 0 {
			// the agent will always return all the interfaces, both the
			// ones managed by libvirt and the ones attached to bridge interfaces
			// or macvtap. Hence it has the highest priority
			return interfaces, nil
		}
	} else {
		log.Println("[DEBUG] skipping qemu-agent")
	}

	// ...
}
```

Then, we could run our tests with `go test -ldflags "-X github.com/dmacvicar/terraform-provider-libvirt/libvirt.skipQemuAgent=true" -v -covermode=count -coverprofile=profile.cov -timeout=1200s ./libvirt`, causing the test cases to always fall back on the libvirt API.

This was the first workaround which came to my mind so it might not be the best. What do you think?

Reply:
You are right. Tests are timing out because we are insisting on trying connect to the `qemu-agent`. I like your idea of having a _qemu-agent switch_, but I would use an environment variable (ie, `TF_SKIP_QEMU_AGENT`) instead of a global because it seems more natural to me (we already have some other `TF_` vars)